### ISSUE TITLE
Add BiDi support

### ISSUE THREAD

**Issue by ahangarha** [Chocobozzz/PeerTube#2113]
**What happened?**
Peertube is LTR friendly. So if you post something in RTL scripts, it behaves wrongly

**What do you expect to happen instead?**
SImple: it should be both LTR and RTL friendly.

**Steps to reproduce:**
Just post something in Persian/Arabic/Urdu etc. Especially it is horrible if it be a mixed LTR and RTL in the text. you may check [this post](https://peertube.social/videos/watch/5c9cb7bb-48a2-4238-acf2-d81a6045d17e) to see an example

**Additional information**
* PeerTube version or URL: PeerTube v1.4.0 on peertube.social
* Browser name/version: Firefox 69 (it is not related to browser version)
* NodeJS version: -

My suggestion is to add bidi support to the UI. This way, regardless of the default text direction, texts will find the right direction based on the first character of the line/paragraph.

I myself have experience of adding bidi support. So you may assign this to me.

**Comment by Chocobozzz** [comment]
I'm sorry I don't know RTL languages, so I'm not sure what I have to do here. Do you have some hints or resources on how we could add RTL support in PeerTube? Or maybe you have already started such work?

**Comment by ahangarha** [comment]
The simplest way is to add `dir="auto"` attribute to all the elements may contain texts which might be in different languages. I haven't checked the source code, otherwise I myself would suggest to give the task to me.

If needed, I can check for it

**Comment by ahangarha** [comment]
I have made a very simple Firefox extension called [Add Bidi Support](https://addons.mozilla.org/en/firefox/addon/add-bidi-support/) which does what I proposed here.

I share some screenshot of how effectively it solves the problem:

![image](https://user-images.githubusercontent.com/11241315/82123834-acf50b00-97b0-11ea-9bf2-8ed3492bfabf.png)

![image](https://user-images.githubusercontent.com/11241315/82123864-ea599880-97b0-11ea-94aa-a4f38c390ae6.png)

![image](https://user-images.githubusercontent.com/11241315/82123927-445a5e00-97b1-11ea-8f18-fc72488147d6.png)

**Comment by ahangarha** [comment]
Any update in this regard?

**Comment by Chocobozzz** [comment]
Hello,

Not yet but I want to work on this this year adding the `dir` attribute to the `body` tag, and using custom SASS functions to set the appropriate `margin-right`/`margin-left` to elements depending on the direction (similar to what bootstrap 5 does). Since it's the first time I'll work on this kind of subject, it will take some time :)

**Comment by ahangarha** [comment]
Adding `dir` to `body` wont solve the issue. Regardless of the overall direction of the page (which should be dependent on UI language), there are mixed content on Fediverse. So it should be handled on each entry level.

Moreover I would suggest to use `margin-inline` and `padding-inline` with `start|end` instead of `left|right` to address bidirectional functionality. That includes `text-align` as well.

**Comment by Chocobozzz** [comment]
> So it should be handled on each entry level

I'm not sure to understand. Are you explaining that we should detect the language in a HTML element to know if it should be displayed to the according direction?

**Comment by ahangarha** [comment]
Yes. If you see the screenshot I shared above, in a timeline or search result or even in a channel, one may have videos with both English and Persian text. English titles should render with LTR direction and Persian titles should render with RTL. This applies to dscription, tags, channel name, etc. This mean each element that contains text which can be either RTL or LTR, should determine the direction of the text and accordingly render it in right direction.

There are different ways to achieve this. One is to apply `dir="auto"` to the element. The other is to add `unicode-bidi: plaintext` style to the element which works but is not recommended. The first approach is used by many projects including Gitlab. In either cases, we should avoid using absolute values of `left|right` for text alignment, padding and margin and border unless.

**Comment by ahangarha** [comment]
See. Peertube (like any other social media platform or messengers) hosts videos from different languages. Channel name, Video title, video description, tags, etc can not only have text in different languages but also can have mixed text. Not all people see videos on peertube while being logged in. So they cannot choose language. Even if they do this, they see problems in text direction specially if there be a mixed text.

Having RTL support is good but it is different from supporting Bidirectional text. If you share your concern regarding layout, I might be able to help to fix them in the best possible way.

Social media is not where we can control the language people share content. The platform should be enough flexible to handle it. Trust me! I along with millions of other people are suffering from this issue. So lets work together and fix this issue. It might take some time but it should be resolved.

Since the issue is not resolved, I request to keep it open. The mentioned commit 27bc958 doesn't address the issue raised here. it is RTL support not bidi support.

**Comment by ahangarha** [comment]
Thank you @Chocobozzz. I really appreciate it.

**Comment by jermanuts** [comment]
can you check now @ahangarha ? Peertube have improved RTL a lot recently

**Comment by ahangarha** [comment]
Hi @jermanuts,

I checked TubEdu (using version 8.2.2) and I see at least video titles having problem. See the example below:

<img width="824" height="247" alt="Image" src="https://github.com/user-attachments/assets/7614e56a-f77b-4d01-8be3-27480ace424a" />

[link to the video](https://tubedu.org/w/m7Xeud3DD7t5g6wdM5FRiG)

Whenever you are ready, let me know so that we can work on issues. If I prepare the dev environment, I can start makeing PRs. But for that, we need good level of cooperation to ensure everything goes right.

Also if you want to work on it yourself, let me know so that I can provide help.

**Comment by jermanuts** [comment]
If you change language in settings to Arabic or any other RTL language, this is set correctly to RTL.

I think what you want is detecting RTL text and automatically set it to RTL even on LTR language interfaces (like in Title, description and comments etc) and vice versa for RTL interface with LTR text. Which is nice and easy I think for @Chocobozzz . Youtube themselves didn't implement it for some reason.

But for profile pictures like the video author and commenters should stay in LTR position, also their text? I guess here where problem arises because of content shifting based on the video.

**Comment by ahangarha** [comment]
There are two points:
- The key issue is not to add RTL/LTR support, but to add support bidirectional text. This means user generated content to be shown in correct direction regardless of the parents direction. If we don't do this, users cannot read mixed content when the direction is wrong.
- I recognize your concern over the layout design as a valid concern. But here we need to distinguish direction from alignment. I think what you care here is to enforce text alignment based on the global direction so if the language is set to English, all texts should be left aligned, and in the same way, if user chooses Persian/Arabic/... the text alignment must be right by force.

New visual examples:

<img width="1123" height="174" alt="Image" src="https://github.com/user-attachments/assets/eb1da3b6-82dd-4a85-a9c7-75324d140d7e" />

<img width="644" height="241" alt="Image" src="https://github.com/user-attachments/assets/35b69e13-ce8d-49f5-9592-1fa99ed81530" />

In the above examples, I tried to demonstrate the different scenarios.
- The red line: direction is wrong so the order of words (read from right to left) is LTR (following the context)
- The violet line has correct direction but alignment (which is dynamic and follows the direction because of `text-align: start`) breaks the layout design.
- The yellow line has correct direction (based on content not context) but alignment is set based on the layout direction (`text-align: left` because the layout is LTR).

I think last option is what addresses both issues correctly.

I hope I could make the issue more clear now.

---

In case you have issue with understanding the order of the words, the following lines may help. I think they make it easy to verity the correct direction (order) of the text. Use them as the text where ever possibly we can have text in either of the directions:

یک two سه four پنج six هفت eight در RTL

one دو three چهار five شش seven در LTR نه

**Comment by Chocobozzz** [comment]
Thank you for the explanations @ahangarha, they are very interesting.

Can you propose a PR for the yellow line, just for the video watch page so we can see how to handle BiDi technically?

**Comment by jermanuts** [comment]
>The violet line has correct direction but alignment (which is dynamic and follows the direction because of text-align: start) breaks the layout design.

This is the correct way, the yellow way is wrong, mixed languages is read based on main language used. IF it is arabic you will always be reading from the right to left even when you reach LTR text, then you will LTR text from the left to right until you reach RTL text where you continue reading from Right to left

I can't see how yellow way would work in a large paragraph with mixed languages RTL and LTR. your example was 1 liner so it may not been clear what you meant.

**Comment by ahangarha** [comment]
If I can run the project locally, I send a PR. We can take as the starting point. It would be easier and more realistic to discuss on actual result.

Currently I am busy with some stuff. But I try to make the PR before August.
